tap_zendesk_chat/__init__.py

```diff
@@ -1,134 +1,25 @@
-        else:
-            raise
-    return True
-
-
-def discover(config):
-    client = Client(config)
-    ensure_credentials_are_authorized(client)
-    include_account_stream = is_account_endpoint_authorized(client)
-    catalog = Catalog([])
-    for stream in streams_.all_streams:
-        if (not include_account_stream
-            and stream.tap_stream_id == streams_.ACCOUNT.tap_stream_id):
-            continue
-        raw_schema = load_schema(stream.tap_stream_id)
-        mdata = build_metadata(raw_schema, stream)
-        schema = Schema.from_dict(raw_schema)
-        catalog.streams.append(CatalogEntry(
-            stream=stream.tap_stream_id,
-            tap_stream_id=stream.tap_stream_id,
-            key_properties=stream.pk_fields,
-            schema=schema,
-            metadata=metadata.to_list(mdata)
-        ))
-    return catalog
-
-def build_metadata(raw_schema, stream):
-
-    mdata = metadata.new()
-    metadata.write(mdata, (), 'valid-replication-keys', list(stream.replication_key))
-    metadata.write(mdata, (), 'table-key-properties', list(stream.pk_fields))
-    for prop in raw_schema['properties'].keys():
-        if prop in stream.replication_key or prop in stream.pk_fields:
-            metadata.write(mdata, ('properties', prop), 'inclusion', 'automatic')
-        else:
-            metadata.write(mdata, ('properties', prop), 'inclusion', 'available')
-
-    return mdata
-
-
-def output_schema(stream):
-    schema = load_schema(stream.tap_stream_id)
-    singer.write_schema(stream.tap_stream_id, schema, stream.pk_fields)
-
-
-def is_selected(stream):
-    mdata = metadata.to_map(stream.metadata)
-    return metadata.get(mdata, (), 'selected')
-
-def sync(ctx):
-    currently_syncing = ctx.state.get("currently_syncing")
-    start_idx = streams_.all_stream_ids.index(currently_syncing) \
-        if currently_syncing else 0
-    stream_ids_to_sync = [cs.tap_stream_id for cs in ctx.catalog.streams
-                          if is_selected(cs)]
-    streams = [s for s in streams_.all_streams[start_idx:]
-               if s.tap_stream_id in stream_ids_to_sync]
-    for stream in streams:
-        ctx.state["currently_syncing"] = stream.tap_stream_id
-        output_schema(stream)
-        ctx.write_state()
-        stream.sync(ctx)
-    ctx.state["currently_syncing"] = None
-    ctx.write_state()
-
-
-def main_impl():
-    args = utils.parse_args(REQUIRED_CONFIG_KEYS)
+@handle_top_exception(LOGGER)
+def main():
+    """performs sync and discovery."""
+    args = parse_args(REQUIRED_CONFIG_KEYS)
     if args.discover:
         discover(args.config).dump()
-        print()
     else:
-        catalog = Catalog.from_dict(args.properties) \
-            if args.properties else discover(args.config)
-        ctx = Context(args.config, args.state, catalog)
-        sync(ctx)
+        ctx = Context(args.config, args.state, args.catalog or discover(args.config))
+        sync(ctx, args.catalog or discover(args.config))
```

Comment:
Why is args.catalog being consumed in both the lines? ctx would have catalog attribute

Reply:
fixed